drivers/serial/uart_smartbond.c

```diff
@@ -238,11 +311,104 @@ static int uart_smartbond_config_get(const struct device *dev,
+					uint32_t pins)
+{
+	struct uart_smartbond_data *data = CONTAINER_OF(cb, struct uart_smartbond_data,
+							rx_wake_cb);
+
+	/* Disable interrupts on UART RX pin to avoid repeated interrupts. */
+	(void)gpio_pin_interrupt_configure(gpio, (find_msb_set(pins) - 1),
+					   GPIO_INT_DISABLE);
+	/* Refresh console expired time */
+	if (data->rx_wake_timeout) {
+
+		uart_smartbond_pm_policy_state_lock_get(data,
+							UART_SMARTBOND_PM_POLICY_STATE_RX_FLAG);
+		k_work_reschedule(&data->rx_timeout_work, K_MSEC(data->rx_wake_timeout));
+	}
+}
+
+static void uart_smartbond_dtr_handler(const struct device *gpio, struct gpio_callback *cb,
+				       uint32_t pins)
+{
+	struct uart_smartbond_data *data = CONTAINER_OF(cb, struct uart_smartbond_data,
+							dtr_wake_cb);
+	int pin = find_lsb_set(pins) - 1;
+
+	if (gpio_pin_get(gpio, pin) == 1) {
+		uart_smartbond_pm_policy_state_lock_put(data,
+							UART_SMARTBOND_PM_POLICY_STATE_DTR_FLAG);
+	} else {
+		uart_smartbond_pm_policy_state_lock_get(data,
+							UART_SMARTBOND_PM_POLICY_STATE_DTR_FLAG);
+	}
+}
+
+#endif
+
 static int uart_smartbond_init(const struct device *dev)
 {
 	struct uart_smartbond_data *data = dev->data;
+	int ret = 0;
+
+	da1469x_pd_acquire(MCU_PD_DOMAIN_COM);
+
+#ifdef CONFIG_PM_DEVICE
+	int rx_wake_timeout;
+	const struct uart_smartbond_cfg *config = dev->config;
+	const struct device *uart_console_dev =
+		DEVICE_DT_GET(DT_CHOSEN(zephyr_console));
+	/* All uarts can have wake time specified in device tree to keep
+	 * device awake after receiving data
+	 */
+	rx_wake_timeout = config->rx_wake_timeout;
+	if (dev == uart_console_dev) {
+#ifdef CONFIG_UART_CONSOLE_INPUT_EXPIRED
+		/* For device configured as console wake time is taken from
+		 * Kconfig same way it is configured for other platforms
+		 */
+		rx_wake_timeout = CONFIG_UART_CONSOLE_INPUT_EXPIRED_TIMEOUT;
+#endif
+	}
+	/* If DTR pin is configured, use it for power management */
+	if (config->dtr_gpio.port != NULL) {
+		gpio_init_callback(&data->dtr_wake_cb, uart_smartbond_dtr_handler,
+				   BIT(config->dtr_gpio.pin));
+		ret = gpio_add_callback(config->dtr_gpio.port, &data->dtr_wake_cb);
+		if (ret == 0) {
+			ret = gpio_pin_interrupt_configure_dt(&config->dtr_gpio,
+							      GPIO_INT_MODE_EDGE |
+							      GPIO_INT_TRIG_BOTH);
+			/* Check if DTR is already active (low), if so lock power state */
+			if (gpio_pin_get(config->dtr_gpio.port, config->dtr_gpio.pin) == 0) {
+				uart_smartbond_pm_policy_state_lock_get(data,
+					UART_SMARTBOND_PM_POLICY_STATE_DTR_FLAG);
+			}
+		}
+	}
+	if (rx_wake_timeout > 0 && config->rx_wake_gpio.port != NULL) {
+		k_work_init_delayable(&data->rx_timeout_work,
+				      uart_smartbond_rx_refresh_timeout);
+		gpio_init_callback(&data->rx_wake_cb, uart_smartbond_wake_handler,
+				   BIT(config->rx_wake_gpio.pin));
+
+		ret = gpio_add_callback(config->rx_wake_gpio.port, &data->rx_wake_cb);
+		if (ret < 0) {
+			return ret;
```

Comment:
Should we call `da1469x_pd_release(MCU_PD_DOMAIN_COM);` before returning?

Reply:
@ydamigos you are right, code will be changed not to return here. If add callback fails (unlikely) timeout will not be configured and UART will work except timeout functionality will be inactive